spring-cloud-dataflow-completion/src/main/java/org/springframework/cloud/dataflow/completion/RecoveryStrategy.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.dataflow.completion;
+
+import java.util.List;
+
+/**
+ * Used to provide completions on ill-formed stream definitions, after an initial (failed) parse.
+ * 
+ * @param <E> the kind of exception that is intercepted during parsing
+ * 
+ * @author Eric Bottard
+ */
+public interface RecoveryStrategy<E extends Exception> {
```

Comment:
No, as we need to decide via stacktrace fingerprinting whether or not to trigger.
Moreover, I'm not that happy with the change we did in ExpansionStrategy. For some implementations it's ok, for others there is quite a cognitive burden at the start of the method to decide not to do anything